drivers/serial/uart_nrf_sw_lpuart.c

```diff
@@ -0,0 +1,658 @@
+				data->rx_len, data->rx_timeout);
+	__ASSERT(err == 0, "RX: Enabling failed (err:%d)", err);
+
+	/* Ready. Confirm by toggling the pin. */
+	ctrl_pin_clear(&data->rdy_pin);
+	ctrl_pin_set(&data->rdy_pin, false);
+	LOG_DBG("RX: Ready");
+	data->rx_req = false;
+	data->rx_state = RX_ACTIVE;
+}
+
+/* Called when end of transfer is detected. It sets response pin to idle and
+ * disables RX.
+ */
+static void deactivate_rx(struct lpuart_data *data)
+{
+	int err;
+
+	ctrl_pin_idle(&data->rdy_pin);
+	if (nrf_gpio_pin_read(data->rdy_pin.nrf_pin)) {
+		LOG_DBG("RX: Request pending while deactivating");
+		/* pin is set high, another request pending. */
+		nrf_gpiote_event_clear(NRF_GPIOTE,
+				     nrf_gpiote_in_event_get(data->rdy_pin.ch));
+		data->rx_req = true;
+	}
+
+	/* abort rx */
+	data->rx_state = RX_TO_IDLE;
+	err = uart_rx_disable(data->uart);
+	if (err < 0) {
+		/* todo notify error, which */
+		LOG_ERR("RX: Failed to disable (err: %d)", err);
+	}
+}
+
+static void tx_complete(struct lpuart_data *data)
+{
+	ctrl_pin_idle(&data->req_pin);
+	data->tx_buf = NULL;
+	data->tx_active = false;
+}
+
+static void on_req_pin_change(struct lpuart_data *data)
+{
+	int key;
+	const uint8_t *buf;
+	size_t len;
+	int err;
+
+	if (data->tx_buf == NULL) {
+		LOG_WRN("TX: request confirmed but no data to send");
+		tx_complete(data);
+		/* aborted */
+		return;
+	}
+
+	LOG_DBG("TX: Confirmed, starting.");
+	ctrl_pin_set(&data->req_pin, true);
+	k_timer_stop(&data->tx_timer);
+
+	key = irq_lock();
+	data->tx_active = true;
+	buf = data->tx_buf;
+	len = data->tx_len;
+	irq_unlock(key);
+	err = uart_tx(data->uart, buf, len, 0);
+	if (err < 0) {
+		LOG_ERR("TX: Not started (error: %d)", err);
+		tx_complete(data);
+	}
+}
+
+static void on_rdy_pin_change(struct lpuart_data *data)
+{
+	if (nrf_gpiote_event_polarity_get(NRF_GPIOTE, data->rdy_pin.ch)
+		== NRF_GPIOTE_POLARITY_LOTOHI) {
+		__ASSERT_NO_MSG(data->rx_state != RX_ACTIVE);
+
+		LOG_DBG("RX: Request detected.");
+		data->rx_req = true;
+		if (data->rx_state == RX_IDLE) {
+			activate_rx(data);
+		}
+	} else {
+		__ASSERT_NO_MSG(data->rx_state == RX_ACTIVE);
+
+		LOG_DBG("RX: End detected.");
+		deactivate_rx(data);
+	}
+}
+
+static void gpio_handler(struct device *port,
+			struct gpio_callback *cb,
+			gpio_port_pins_t pins)
+{
+	struct lpuart_bidir_gpio *io =
+		CONTAINER_OF(cb, struct lpuart_bidir_gpio, callback);
+	struct lpuart_data *data;
+
+	if (io->req) {
+		data = CONTAINER_OF(io, struct lpuart_data, req_pin);
+		on_req_pin_change(data);
+		return;
+	}
+
+	data = CONTAINER_OF(io, struct lpuart_data, rdy_pin);
+	on_rdy_pin_change(data);
+}
+
+static int ctrl_pin_configure(struct lpuart_bidir_gpio *io, struct device *port,
+				int pin, bool req)
+{
+	int err;
+
+	io->callback.handler = gpio_handler;
+	io->callback.pin_mask = BIT(pin);
+	io->pin = pin;
+	io->port = port;
+	io->req = req;
+	io->nrf_pin = pin +
+		((io->port == device_get_binding("GPIO_0")) ? 0 : 32);
```

Comment:
Couldn't it be made simpler, without calling `device_get_binding()`? You've got exactly these values for both pins in DT node properties.

Reply:
reworked so that function accepts const configuration structure which contains pin and nrf_pin.